### ISSUE TITLE
AsyncFile.read allocate two buffer to read the data

### ISSUE THREAD

**Issue by luengnat** [eclipse-vertx/vert.x#2511]
AsyncFile.read allocate two buffers to read the data makes it very inefficient.

**Comment by vietj** [comment]
can you point out which part exactly ?

**Comment by luengnat** [comment]
Once here https://github.com/eclipse/vert.x/blob/6a01366c8b86f4d4431d9e97d775c45628dcd0ce/src/main/java/io/vertx/core/file/impl/AsyncFileImpl.java#L323

and another at 
https://github.com/eclipse/vert.x/blob/6a01366c8b86f4d4431d9e97d775c45628dcd0ce/src/main/java/io/vertx/core/file/impl/AsyncFileImpl.java#L130

The buffer seems to be allocated for read and then use to copy into writeBuff here 
https://github.com/eclipse/vert.x/blob/6a01366c8b86f4d4431d9e97d775c45628dcd0ce/src/main/java/io/vertx/core/file/impl/AsyncFileImpl.java#L424 
and then get thrown away.

**Comment by vietj** [comment]
the two buffers are different, one is a user buffer that will contain the result and the other is a temporary java `ByteBuffer`. I think it can be complicated to reuse the same buffer because it will require to use an internal Netty API, however I think that in the situation where the user sets an handler on the `AsyncFile` then we can allocate a single `ByteBuffer` and reuse it for read operations, that would not occur for a single read.

**Comment by luengnat** [comment]
Yeah. You're right. At least, you should be able to optimize such that https://github.com/eclipse/vert.x/blob/6a01366c8b86f4d4431d9e97d775c45628dcd0ce/src/main/java/io/vertx/core/file/impl/AsyncFileImpl.java#L130 is allocated just once and it can be reused for the duration of the repeated read right?

**Comment by vietj** [comment]
yes, it could be allocated once before and passed during trampolining

**Comment by tsegismont** [comment]
Isn't it implemented like this already?

https://github.com/eclipse/vert.x/blob/6a01366c8b86f4d4431d9e97d775c45628dcd0ce/src/main/java/io/vertx/core/file/impl/AsyncFileImpl.java#L438

**Comment by vietj** [comment]
that's only for a single chunk read that retries to get the remaining of the buffer to read

**Comment by luengnat** [comment]
https://github.com/eclipse/vert.x/blob/6a01366c8b86f4d4431d9e97d775c45628dcd0ce/src/main/java/io/vertx/core/file/impl/AsyncFileImpl.java#L123 is called repeatedly for each individual chunk. Therefore, ByteBuffer.allocate is called over and over again by doRead() method. 

ByteBuffer should be safe to reuse but Buffer is not safe to reuse.

**Comment by tsegismont** [comment]
Oh right, when used as `ReadStream`. Thanks

**Comment by luengnat** [comment]
@tsegismont That's right.

**Comment by tsegismont** [comment]
@luengnat would you like to contribute this enhancement?

>

**Comment by KowalczykBartek** [comment]
I have a question about this API usage, how users mostly use 
```java
AsyncFile read(Buffer buffer, int offset, long position, int length, Handler<AsyncResult<Buffer>> handler);
```
? do they like in **testReadAsync** test ? 
```java
@Test
public void testReadAsync() throws Exception {
  String fileName = "some-file.dat";
  int chunkSize = 1000;
  int chunks = 10;
  byte[] content = TestUtils.randomByteArray(chunkSize * chunks);
  Buffer expected = Buffer.buffer(content);
  createFile(fileName, content);
  AtomicInteger reads = new AtomicInteger();
  vertx.fileSystem().open(testDir + pathSep + fileName, new OpenOptions(), arr -> {
    if (arr.succeeded()) {
      Buffer buff = Buffer.buffer(chunks * chunkSize);
      for (int i = 0; i < chunks; i++) {
        arr.result().read(buff, i * chunkSize, i * chunkSize, chunkSize, arb -> {
          if (arb.succeeded()) {
            if (reads.incrementAndGet() == chunks) {
              arr.result().close(ar -> {
                if (ar.failed()) {
                  fail(ar.cause().getMessage());
                } else {
                  assertEquals(expected, buff);
                  assertEquals(buff, arb.result());
                  testComplete();
                }
              });
            }
          } else {
            fail(arb.cause().getMessage());
          }
        });
      }
    } else {
      fail(arr.cause().getMessage());
    }
  });
  await();
}
```
I mean, schedule one-by-one some reads and then make usage of results, if yes, usage of this new signature (with reusable ByteBuffer) can be error-prone, just look at my test (https://github.com/KowalczykBartek/vert.x/commit/90b1cb37c6646d8ff8f0f83fdea42da66ed65b9c#diff-19eb29ceb17c93495465a4e22b41eca3R1101), because ByteBuffer is accessed by multiple threads (by SimpleAsynchronousFileChannelImpl executor), such read often fails.

**Comment by vietj** [comment]
i've closed this issue because it won't be fixed however the discussion lead to open this issue https://github.com/eclipse-vertx/vert.x/issues/2732 that is fixed in 3.6.0
